Common/src/main/java/com/blamejared/crafttweaker/api/tag/manager/ITagManager.java

```diff
@@ -1,187 +1,158 @@
 import net.minecraft.resources.ResourceLocation;
 import net.minecraft.tags.Tag;
-import net.minecraft.tags.TagCollection;
+import net.minecraft.tags.TagManager;
 import org.openzen.zencode.java.ZenCodeType;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
+import java.util.Objects;
 import java.util.stream.Collectors;
 
-/**
- * TagManagers are used to handle the different types of Tags within the game.
- * They can be retrieved directly with the ITagManager BEP, and are also used indirectly when creating a tag with the Tag BEP.
- *
- * @param <T> The element types of this tag.
- */
 @ZenRegister
-@Document("vanilla/api/tag/ITagManager")
-@ZenCodeType.Name("crafttweaker.api.tag.ITagManager")
-public interface ITagManager<T> extends CommandStringDisplayable {
-    
-    @Nonnull
-    Class<T> getElementClass();
-    
-    /**
-     * Retrieves a tag by its name.
-     * Will also be called by the BEP.
-     * <p>
-     * Note that this method does _not_ yet create the tag if it does not exist.
-     * Adding something to the object created by this tag will create it for the game.
-     *
-     * @param name The Resource location of the tag
-     *
-     * @return A Tag object.
-     */
+@Document("vanilla/api/tag/manager/ITagManager")
+@ZenCodeType.Name("crafttweaker.api.tag.manager.ITagManager")
+public interface ITagManager<T> extends CommandStringDisplayable, Iterable<MCTag<T>> {
+
+    @ZenCodeType.Method
+    @ZenCodeType.Getter("tagFolder")
+    default String tagFolder() {
+
+        String tagDir = TagManager.getTagDir(resourceKey());
+
+        // Really not ideal, but I don't see a better way, lets just hope that other mods don't be dumb and add their tags to other folders.
+        if (tagDir.startsWith("tags/")) {
+            tagDir = tagDir.substring("tags/".length());
+        }
+        return tagDir;
+    }
+
     @ZenCodeType.Method
-    default MCTag<T> getTag(String name) {
-        
-        return getTag(new ResourceLocation(name));
+    default MCTag<T> tag(String id) {
+
+        ResourceLocation rl = ResourceLocation.tryParse(id);
+        Objects.requireNonNull(rl, "Invalid Resource Location passed! Given: '" + id + "'");
+        return tag(rl);
     }
-    
-    /**
-     * Retrieves a tag by its name.
-     * Will also be called by the BEP.
-     * <p>
-     * Note that this method does _not_ yet create the tag if it does not exist.
-     * Adding something to the object created by this tag will create it for the game.
-     *
-     * @param location The Resource location of the tag
-     *
-     * @return A Tag object.
-     */
+
     @ZenCodeType.Method
-    default MCTag<T> getTag(ResourceLocation location) {
-        
-        return new MCTag<>(location, this);
+    default MCTag<T> tag(ResourceLocation id) {
+
+        return tagMap().getOrDefault(id, new MCTag<>(id, this));
```

Comment:
Is there a way to avoid creating a new tag for every invocation of this method that does not over-complicate it?

**Only an idea:** why not `computeIfAbsent` and then add the tag as it gets referenced? Could this lead to too much tag bloat?

Reply:
Adding a tag as you call it would make:
```
<tag:items:foo:bar>.exists
```
always return true